src/main/java/io/supertokens/session/accessToken/AccessTokenSigningKey.java

```diff
@@ -69,142 +73,206 @@ public static AccessTokenSigningKey getInstance(Main main) {
         // have to upgrade from the readLock to a
         // writeLock - which is not possible:
         // https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html
-        if (this.keyInfo == oldKeyInfo) {
+
+        // This reference comparison should work, since we recreate the list object each time we refresh and it's unmodifiable
+        if (this.validKeys == oldKeyInfo) {
             // key has not changed since we previously tried to use it.. So we can make it null.
             // otherwise we might end up making this null unnecessarily.
 
             ProcessState.getInstance(this.main)
                     .addState(ProcessState.PROCESS_STATE.SETTING_ACCESS_TOKEN_SIGNING_KEY_TO_NULL, null);
-            this.keyInfo = null;
+            this.validKeys = null;
         }
     }
 
-    public synchronized AccessTokenSigningKey.KeyInfo getKey()
+    public synchronized List<KeyInfo> getKey()
             throws StorageQueryException, StorageTransactionLogicException {
-        if (this.keyInfo == null) {
-            this.keyInfo = maybeGenerateNewKeyAndUpdateInDb();
+        if (this.validKeys == null || this.validKeys.size() == 0) {
+            this.validKeys = maybeGenerateNewKeyAndUpdateInDb();
         }
 
         CoreConfig config = Config.getConfig(main);
-        if (config.getAccessTokenSigningKeyDynamic() && System.currentTimeMillis() > (this.keyInfo.createdAtTime
+        if (config.getAccessTokenSigningKeyDynamic() && System.currentTimeMillis() > (this.validKeys.get(0).createdAtTime
                 + config.getAccessTokenSigningKeyUpdateInterval())) {
             // key has expired, we need to change it.
-            this.keyInfo = maybeGenerateNewKeyAndUpdateInDb();
+            this.validKeys = maybeGenerateNewKeyAndUpdateInDb();
         }
-        return this.keyInfo;
+        return this.validKeys;
+    }
+
+    public KeyInfo getCurrentKey()
+            throws StorageQueryException, StorageTransactionLogicException {
+        return this.getKey().get(0);
     }
 
     public synchronized long getKeyExpiryTime() throws StorageQueryException, StorageTransactionLogicException {
         if (Config.getConfig(this.main).getAccessTokenSigningKeyDynamic()) {
             this.getKey();
-            long createdAtTime = this.keyInfo.createdAtTime;
+            // getKey ensures we have at least 1 valid keys
+            long createdAtTime = this.validKeys.get(0).createdAtTime;
             return createdAtTime + Config.getConfig(main).getAccessTokenSigningKeyUpdateInterval();
         } else {
             // return 10 years from now
             return System.currentTimeMillis() + (10L * 365 * 24 * 3600 * 1000);
         }
     }
 
-    private KeyInfo maybeGenerateNewKeyAndUpdateInDb() throws StorageQueryException, StorageTransactionLogicException {
+    private List<KeyInfo> maybeGenerateNewKeyAndUpdateInDb() throws StorageQueryException, StorageTransactionLogicException {
```

Comment:
I think it may make the logic simpler if we decouple the deleting of keys from this method. We can do this by creating a cronjob to delete older keys, and in this method, we fetch all keys and filter the list to remove the keys that are older than `keysValidAfter`.